ngraph/core/reference/include/ngraph/runtime/reference/gather_nd.hpp

```diff
@@ -26,171 +28,117 @@ namespace ngraph
+                                next(begin(params_shape), batch_dims),
+                                begin(indices_shape)))
                 {
-                    params_outer_end_corner[i] = 1;
+                    throw std::domain_error{
+                        "dimensions in params and indices have to be equal on batch dimensions"};
                 }
-                Strides params_strides(params_ndim, 1);
-                AxisVector params_axis_order(params_ndim);
-                std::iota(params_axis_order.begin(), params_axis_order.end(), 0);
-                CoordinateTransform params_outer_transform(params_shape,
-                                                           params_outer_start_corner,
-                                                           params_outer_end_corner,
-                                                           params_strides,
-                                                           params_axis_order);
-
-                size_t out_ndim = static_cast<size_t>(out_shape.size());
-                Coordinate out_start_corner(out_ndim, 0);
-                Coordinate out_end_corner(out_shape);
-                for (size_t i = 1; i < out_ndim; i++)
+
+                const auto first_slice_index_in_params = batch_dims + indices_shape.back();
+
+                if (!(first_slice_index_in_params <= params_shape.size()))
                 {
-                    out_end_corner[i] = 1;
+                    throw std::domain_error{
+                        "params_shape should have enough rank to be index by indices"};
                 }
-                Strides out_strides(out_ndim, 1);
-                AxisVector out_axis_order(out_ndim);
-                std::iota(out_axis_order.begin(), out_axis_order.end(), 0);
-                CoordinateTransform out_transform(
-                    out_shape, out_start_corner, out_end_corner, out_strides, out_axis_order);
-
-                Shape indices_shape_batch(indices_shape.begin() + batch_dims, indices_shape.end());
-                Shape params_shape_batch(params_shape.begin() + batch_dims, params_shape.end());
-                Shape output_shape_batch(out_shape.begin() + 1, out_shape.end());
-                auto out_coord_iter = out_transform.begin();
-                auto params_coord_iter = params_outer_transform.begin();
-                for (const Coordinate& indices_coord : indices_outer_transform)
+
+                const auto slice_shape = Shape(
+                    next(begin(params_shape), first_slice_index_in_params), end(params_shape));
+                const auto slice_size = shape_size(slice_shape);
+
+                const auto indices_offsets = [&] {
+                    std::vector<size_t> offsets{slice_size};
+                    const auto beg_s = next(rbegin(params_shape), slice_shape.size());
+                    const auto end_s = next(beg_s, indices_shape.back() - 1);
+                    for (auto s = beg_s; s != end_s; ++s)
```

Comment:
If I understand correctly, you're iterating over dimensions here. Thus the `s` variable is misleading and could be replaced with something like `dim`.